source/opt/folding_rules.cpp

```diff
@@ -2760,6 +2760,228 @@ FoldingRule RedundantSUMod() {
+        bool can_remove_inner = true;
+
+        ZipIntegerConstants(
+          const_mgr, const_input1, const_input2,
+          [&can_remove_inner, &can_convert_to_const](auto lhs, auto rhs) {
+            // Only convert to const if 'and' is a subset of 'or'
+            can_convert_to_const = can_convert_to_const && ((lhs & rhs) == lhs);
+            // Only remove 'xor'/'or' if no bits intersect with 'and'
+            can_remove_inner = can_remove_inner && ((lhs & rhs) == 0);
+          });
+
+        if (can_convert_to_const) {
+          Instruction* const_inst =
+            const_mgr->GetDefiningInstruction(const_input1);
+          inst->SetOpcode(spv::Op::OpCopyObject);
+          inst->SetInOperands({{SPV_OPERAND_TYPE_ID, {const_inst->result_id()}}});
+          return true;
+        } else if (can_remove_inner) {
+          Instruction* non_const_input =
+            NonConstInput(context, other_constants[0], other_inst);
+          Instruction* const_inst =
+            const_mgr->GetDefiningInstruction(const_input1);
+          inst->SetInOperands(
+            {{SPV_OPERAND_TYPE_ID, {non_const_input->result_id()}},
+            {SPV_OPERAND_TYPE_ID, {const_inst->result_id()}}});
+          return true;
+        }
+      }
+      return false;
+    };
+}
+
+// Folds redundant add and sub ops that are part of an and.
+// Cases handled:
+// 1 & (b + 2) = b & 1
+// 1 & (b - 2) = b & 1
+FoldingRule RedundantAndAddSub() {
+  return [](IRContext* context, Instruction* inst,
+    const std::vector<const analysis::Constant*>& constants) {
+      assert(inst->opcode() == spv::Op::OpBitwiseAnd && "Wrong opcode.");
+      const analysis::Type* type =
+        context->get_type_mgr()->GetType(inst->type_id());
+      uint32_t width = ElementWidth(type);
+      if ((width != 32) && (width != 64)) return false;
+
+      analysis::ConstantManager* const_mgr = context->get_constant_mgr();
+      const analysis::Constant* const_input1 = ConstInput(constants);
+      if (!const_input1) return false;
+      Instruction* other_inst = NonConstInput(context, constants[0], inst);
+
+      if (other_inst->opcode() == spv::Op::OpIAdd ||
+        other_inst->opcode() == spv::Op::OpISub) {
+        std::vector<const analysis::Constant*> other_constants =
+          const_mgr->GetOperandConstants(other_inst);
+        const analysis::Constant* const_input2 = ConstInput(other_constants);
+        if (!const_input2) return false;
+
+        // Only valid for subtraction if const is on the right
+        if ((other_inst->opcode() == spv::Op::OpISub) && other_constants[0]) {
+          return false;
+        }
+
+        bool can_remove_inner = true;
+        ZipIntegerConstants(const_mgr, const_input1, const_input2,
+          [&can_remove_inner](auto lhs, auto rhs) {
+            can_remove_inner = can_remove_inner &&
+              ((lhs & rhs) == 0) &&
+              (rhs > lhs);
+          });
```

Comment:
I'm not convinced this is correct:

```
0b10 & ( a + 0b101)
```

I believe this would fold to `a & 0b10`.

However, if `a == 0b1` they are different:

```
0b10 & (0b1 + 0b101) = 0b10 & 0b110 = 0b10
0b1 & 0b10 = 0b0
```